src/main/java/com/google/devtools/build/lib/analysis/constraints/RuleContextConstraintSemantics.java

```diff
@@ -808,4 +825,147 @@ private static void addSelectValuesToSet(BuildType.Selector<?> select, final Set
+   * @param violatedConstraint the constraint that the target platform doesn't satisfy. This must be
+   *     null if targetResponsibleForIncompatibility is set.
+   * @throws ActionConflictException if the underlying {@link RuleConfiguredTargetBuilder}
+   *     encounters a problem when assembling a dummy action for the incompatible {@link
+   *     ConfiguredTarget}.
+   */
+  private static ConfiguredTarget createIncompatibleConfiguredTarget(
+      RuleContext ruleContext,
+      ConfiguredTarget targetResponsibleForIncompatibility,
+      ConstraintValueInfo violatedConstraint)
+      throws ActionConflictException {
+    // Create a dummy ConfiguredTarget that has the IncompatiblePlatformProvider set.
+    ImmutableList<Artifact> outputArtifacts = ruleContext.getOutputArtifacts();
+
+    if (ruleContext.isTestTarget() && outputArtifacts.size() == 0) {
+      // Test targets require RunfilesSupport to be added to the RuleConfiguredTargetBuilder
+      // which needs an "executable".  Create one here if none exist already.
+      //
+      // It would be ideal if we could avoid this. Currently the problem is that some rules like
+      // sh_test only declare an output artifact in the corresponding ConfiguredTarget factory
+      // function (see ShBinary.java). Since this code path here replaces the factory function rules
+      // like sh_test never get a chance to declare an output artifact.
+      //
+      // On top of that, the rest of the code base makes the assumption that test targets provide an
+      // instance RunfilesSupport. This can be seen in the TestProvider, the TestActionBuilder, and
+      // the RuleConfiguredTargetBuilder classes. There might be a way to break this assumption, but
+      // it's currently too heavily baked in to work around it more nicely than this.
+      //
+      // Theoretically, this hack shouldn't be an issue because the corresponding actions will never
+      // get executed. They cannot be queried either.
+      outputArtifacts = ImmutableList.of(ruleContext.createOutputArtifact());
+    }
+
+    NestedSet<Artifact> filesToBuild =
+        NestedSetBuilder.<Artifact>stableOrder().addAll(outputArtifacts).build();
+
+    Runfiles.Builder runfilesBuilder =
+        new Runfiles.Builder(
+            ruleContext.getWorkspaceName(),
+            ruleContext.getConfiguration().legacyExternalRunfiles());
+    Runfiles runfiles =
+        runfilesBuilder
+            .addTransitiveArtifacts(filesToBuild)
+            .addRunfiles(ruleContext, RunfilesProvider.DEFAULT_RUNFILES)
+            .build();
+
+    RuleConfiguredTargetBuilder builder = new RuleConfiguredTargetBuilder(ruleContext);
+    builder.setFilesToBuild(filesToBuild);
+
+    if (targetResponsibleForIncompatibility != null) {
+      builder.add(
+          IncompatiblePlatformProvider.class,
+          IncompatiblePlatformProvider.incompatibleDueToTarget(
+              targetResponsibleForIncompatibility));
+    } else {
```

Comment:
I'd check here that violatedConstraint is not null, and add a final else block that throws an IllegalArgumentException stating that both targetResponsible and violatedConstraint are null.

Reply:
Done